deckhand/tests/unit/engine/test_document_layering.py

```diff
@@ -0,0 +1,392 @@
+
+import os
+import yaml
+
+import testtools
+
+from deckhand.engine import layering
+from deckhand import errors
+
+
+class TestDocumentLayering(testtools.TestCase):
+
+    FAKE_YAML_DATA_2_LAYERS = """
+---
+schema: deckhand/LayeringPolicy/v1
+metadata:
+  schema: metadata/Control/v1
+  name: layering-policy
+data:
+  layerOrder:
+    - global
+    - site
+---
+schema: example/Kind/v1
+metadata:
+  schema: metadata/Document/v1
+  name: global-1234
+  labels:
+    key1: value1
+  layeringDefinition:
+    abstract: true
+    layer: global
+data:
+  a:
+    x: 1
+    y: 2
+  c: 9
+---
+schema: example/Kind/v1
+metadata:
+  schema: metadata/Document/v1
+  name: site-1234
+  layeringDefinition:
+    layer: site
+    parentSelector:
+      key1: value1
+    actions:
+      - method: {method}
+        path: {path}
+data:
+  a:
+    x: 7
+    z: 3
+  b: 4"""
+
+    FAKE_YAML_DATA_3_LAYERS = """
+---
+schema: deckhand/LayeringPolicy/v1
+metadata:
+  schema: metadata/Control/v1
+  name: layering-policy
+data:
+  layerOrder:
+    - global
+    - region
+    - site
+---
+schema: example/Kind/v1
+metadata:
+  schema: metadata/Document/v1
+  name: global-1234
+  labels:
+    key1: value1
+  layeringDefinition:
+    abstract: true
+    layer: global
+data:
+  a:
+    x: 1
+    y: 2
+---
+schema: example/Kind/v1
+metadata:
+  schema: metadata/Document/v1
+  name: region-1234
+  labels:
+    key1: value1
+  layeringDefinition:
+    abstract: true
+    layer: region
+    parentSelector:
+      key1: value1
+    actions:
+      - method: replace
+        path: .a
+data:
+  a:
+    z: 3
+---
+schema: example/Kind/v1
+metadata:
+  schema: metadata/Document/v1
+  name: site-1234
+  layeringDefinition:
+    layer: site
+    parentSelector:
+      key1: value1
+    actions:
+      - method: merge
+        path: .
+data:
+  b: 4"""
+
+    def _read_data(self, filename):
+        dir_path = os.path.dirname(os.path.realpath(__file__))
+        test_yaml_path = os.path.abspath(os.path.join(
+            dir_path, os.pardir, 'resources', filename))
+
+        with open(test_yaml_path, 'r') as yaml_file:
+            yaml_data = yaml_file.read()
+        return self._parse_data(yaml_data)
+
+    def _parse_data(self, yaml_data):
+        return [d for d in yaml.safe_load_all(yaml_data)]
+
+    def _test_layering(self, expected, documents):
+        document_layering = layering.DocumentLayering(documents)
+        rendered_data = document_layering.render()
+        self.assertEqual(expected, rendered_data)
+
+
+class TestDocumentLayering2Layers(TestDocumentLayering):
+
+    def test_layering_method_delete(self):
+        expected = [{}, {'b': 4}]
+
+        for idx, path in enumerate(['.', '.a']):
+            documents = self._parse_data(self.FAKE_YAML_DATA_2_LAYERS.format(
+                method='delete', path=path))
+            self._test_layering(expected[idx], documents)
+
+    def test_layering_method_merge(self):
+        expected = [
```

Comment:
Also, why is "Case `path: .a` I would expect `{a: {x: 7, y: 2, z: 3}, c: 9}`" true? Why is "c" included in the result? Why not "b"? If the merge path is ".a" why is "b" being dropped from the site data and "c" being included from the global data?

I'd imagine that should only be possible if:

 ```python

    kwargs = {
                "_GLOBAL_DATA_": {"data": {"a": {"x": 1, "y": 2}, "c": 9}},
                "_SITE_DATA_": {"data": {"a": {"x": 7, "z": 3}, "b": 4}},
                "_SITE_ACTIONS_": {
                    "actions": [{"method": "merge", "path": ".a"}, {"method": "merge", "path": ".c"}, {"method": "delete", "path": ".b"}]}
            }
```

But as you say above, this would be impossible, since if ".c" is not in the child, you expect a validation error. So how do you get `"c": 9` into the child other than doing a "merge" with a path of ".".